source/nodejs/adaptivecards/src/logging/IACProvider.ts

```diff
@@ -0,0 +1,31 @@
+/*
+	The IACProvider interface. All provider implementations must follow this interface
+	in order to be used by the IACLogger.
+*/
+export interface IACProvider {
```

Comment:
It is also debatable whether you need both a Logger and a Provider type. I think you could have a single Logger type, something along those lines:

```typescript
export enum LogLevel {
    Warning,
    Error,
    Verbose,
    Info,
    Event
}

export abstract class Logger {	
    abstract log(
        logLevel: LogLevel,
        event: string,
        eventSourceName: string,
        message: string,
        correlationID?: string,
        valueSet?: object);
}

export class LoggerCollection extends Logger {
    private _loggers: Logger[];

    log(
        logLevel: LogLevel,
        event: string,
        eventSourceName: string,
        message: string,
        correlationID?: string,
        valueSet?: object) {
        for (let logger of this._loggers) {
            logger.log(logLevel, event, eventSourceName, message, correlationID, valueSet);
        }
    }

    constructor(...loggers: Logger[]) {
        super();

        this._loggers = loggers;
    }
}

export class ConsoleLogger extends Logger {
    log(
        logLevel: LogLevel,
        event: string,
        eventSourceName: string,
        message: string,
        correlationID?: string,
        valueSet?: object) {
        // Emit even to console
    }
}

export class AdaptiveCard {
    ...
    static logger: Logger;
    ...
}

// Use a single ConsoleLogger
AdaptiveCard.logger = new ConsoleLogger();

// Use multiple loggers
AdaptiveCard.logger = new LoggerCollection(new ConsoleLogger(), new Ms1DSLogger());
```

Reply:
Thanks David! We are using a separate provider abstraction to isolate the transport from the core logging APIs. that way we dont have re-implement everything when switching providers/transports.

Reply:
wrt abstract/interface - indeed it is debatable and I'd like us to use interfaces with a default concrete implementation if that's ok. Will look at some of the gaps in your implemetnation though functionality/usability wise and get back.

Reply:
My proposal achieves the exact same result (transport vs core logging)

I actually don't think that's ok for the reasons I mentioned. Please articulate the reasons you think it is better to stick with an interface and let's discuss further.

Reply:
Hi David,

In terms of why we decided to have both Logger and Provider classes:

- We expose an API that allows for logging different levels of messages in an intuitive way (logWarn, logError, etc.). However, this is abstracted by the ACLogger class and instead of several log methods a provider must implement, they are only required to log to two. Requiring these in a single Logger interface/abstract class would put unnecessary burden and redundancy on the developer implementing a custom provider.
- Separating ACLogger (what we implement) and ACProvider (what another party may implement) means that we can catch any errors from the ACProvider within the ACLogger class rather than catching them in the existing code.
- There are some parsing calls (e..g, the switch blocks in the renderCardTelemetry() method) that I am working to move out of card-elements.ts and into the ACLogger class. In the future when we add new events/card elements, we can use the ACLogger to carry out any intermediary steps between a log call and the sending of data.

Reply:
Hey Peter, your second argument makes sense and is enough of a justification for having a logger + provider model, thanks for explaining.

That said, it is still debatable whether the provider model should be interface- or abstract class-based. I believe classes should always be used unless an interface is the only option (and in my opinion that happens only when two classes that do not/cannot extend the same base class need to be treated in a generic way).

In the current context, a class works just fine so that's what we should use IMO. Remember we can basically never change an interface (e.g. add members, especially methods) that's already shipped because it would break external code that implements that interface - in Windows, the overuse of interfaces has lead to ISomething followed by ISomething2 followed by ISomethingEx and all of that confusing nonsense. We can however add members to a class without breaking anything.

Reply:
Hey David, like I said interface vs abstract classes is going to be a debate that i deally dont want to get into. Apart from following the ISP, and fleshing out very specifically the Logger interface + provider interfaces resp, it also allows for easy mocking and testability. Also because of the fact that it's so concise, I dont really expect this interface to change much at all. And in teh case that it does the newer versions inmplement the new interface (because you can have multiple inheritance with interfaces) while the older ones continue to use the older interface and dont have access to the new functionality by design. What scenarios are you concerned will break ? Again, it's pretty standard practice to use and define simple interfaces with more elaborate concrete classes and versioning them esp for simple abstractions like logging-- There's existing solutions at play we've used as guidance: 
 https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.ilogger?view=aspnetcore-2.2 
https://github.com/Azure/azure-functions-host/wiki/ILogger

Reply:
Abstract base classes also allow for easy mocking and testability. It's also very common practice to use abstract base classes over interfaces. It would be great if you could provide concrete arguments/examples that outline how using interfaces is actually better in this context, like I have done for classes. Saying that "you don't envision that the problems that definitely can occur when using interfaces will actually happen" doesn't make much sense to me; I'd rather choose an implementation with which the problems can't happen in the first place.